container/common/helpers.go

```diff
@@ -110,9 +118,25 @@ func GetSpec(cgroupPaths map[string]string, machineInfoFactory info.MachineInfoF
 	if ok {
 		if utils.FileExists(memoryRoot) {
 			spec.HasMemory = true
-			spec.Memory.Limit = readUInt64(memoryRoot, "memory.limit_in_bytes")
-			spec.Memory.SwapLimit = readUInt64(memoryRoot, "memory.memsw.limit_in_bytes")
-			spec.Memory.Reservation = readUInt64(memoryRoot, "memory.soft_limit_in_bytes")
+			if cgroups.IsCgroup2UnifiedMode() {
+				if utils.FileExists(path.Join(memoryRoot, "memory.max")) {
+					spec.Memory.Reservation = readUInt64(memoryRoot, "memory.high")
+					spec.Memory.Limit = readUInt64(memoryRoot, "memory.max")
+					spec.Memory.SwapLimit = readUInt64(memoryRoot, "memory.swap.max")
+				} else {
+					limit, err := machine.GetMachineMemoryCapacity()
```

Comment:
Is the cgroup always unlimited when memory.max doesn't exist?  Could it be affected by the parent cgroup?

If at all possible, we should keep the behavior between the two modes the same.

Reply:
good point, yes we should definitely look into the parent directory and see if any limit is set there.  I've changed it in the new version I've pushed

Reply:
Sorry, that was more of a question than an assertion about the file not existing...  I wasn't sure under what conditions memory.max wouldn't exist.  I took a look at the v2 documentation for it, and the file should exist for non-root cgroups: https://www.kernel.org/doc/Documentation/cgroup-v2.txt.  We just need to handle the case where it is the root cgroup, and use the machine limit as you did before.  I don't think we need to search the higherarchy, since that maintains the same behavior as cgroups v1.

I looked at the cgroup v1 implementation, and it looks like the default value is a very high integer.  For me, it was 9223372036854771712 bytes instead of the machine limit.

Reply:
with cgroups v2 it is possible though to enable the memory controller only for the parent cgroup, where `memory.max` exists and not enable it for the child directory.  In that case the limit still exists.

```
mkdir -p /sys/fs/cgroup/foo/bar
# ls /sys/fs/cgroup/foo/memory.max /sys/fs/cgroup/foo/bar/memory.max
ls: cannot access '/sys/fs/cgroup/foo/bar/memory.max': No such file or directory
/sys/fs/cgroup/foo/memory.max
```

In this case processes added to  `/sys/fs/cgroup/foo/bar` are still limited in memory by   '/sys/fs/cgroup/foo/memory.max`.

Reply:
oh, ok.  Makes sense.